Support connecting to ZooKeeper with TLS

mimaison opened the pull request:
This PR resolves #1323.

Co-authored-by: Mickael Maison <mickael.maison@gmail.com>
Co-authored-by: Edoardo Comar <ecomar@uk.ibm.com>

efeg reviewed (commented):
@mimaison Thanks for the PR.
Looks like there are some style failures -- would you mind resolving them when you get a chance?

```
[ant:checkstyle] [ERROR] /home/circleci/workspace/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlUtils.java:514: Line is longer than 150 characters (found 170). Try to keep lines under 120 characters. [LineLength]
[ant:checkstyle] [ERROR] /home/circleci/workspace/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:13:1: Extra separation in import group before 'java.util.HashMap' [CustomImportOrder]
[ant:checkstyle] [ERROR] /home/circleci/workspace/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:17:1: Utility classes should not have a public or default constructor. [HideUtilityClassConstructor]
[ant:checkstyle] [ERROR] /home/circleci/workspace/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:21:46: Name 'zkProperties' must match pattern '^(_?log)|([A-Z][A-Z0-9]*(_[A-Z0-9]+)*)$'. [ConstantName]
[ant:checkstyle] [ERROR] /home/circleci/workspace/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:38:5: Missing a Javadoc comment. [MissingJavadocMethod]
[ant:checkstyle] [ERROR] /home/circleci/workspace/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:67:21: 'if' construct must use '{}'s. [NeedBraces]
[ant:checkstyle] [ERROR] /home/circleci/workspace/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:72:21: 'if' construct must use '{}'s. [NeedBraces]
[ant:checkstyle] [ERROR] /home/circleci/workspace/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:77:21: 'if' construct must use '{}'s. [NeedBraces]
[ant:checkstyle] [ERROR] /home/circleci/workspace/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:82:21: 'if' construct must use '{}'s. [NeedBraces]
[ant:checkstyle] [ERROR] /home/circleci/workspace/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:85:17: Fall through from the last branch of the switch statement. [FallThrough]
```

egyedt reviewed (changes requested):
Thanks for the PR.
I would recommend to write two integration tests to see if ssl enabled and disabled ZK also works well with the new changes.

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:
```diff
@@ -0,0 +1,103 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(ZKConfigUtils.class);
+
+    private static final Map<String, String> ZK_PROPERTIES = new HashMap<>();
+    static {
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_CNXN_SOCKET_CONFIG, ZKClientConfig.ZOOKEEPER_CLIENT_CNXN_SOCKET);
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_KEYSTORE_LOCATION_CONFIG, "zookeeper.ssl.keyStore.location");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_KEYSTORE_PASSWORD_CONFIG, "zookeeper.ssl.keyStore.password");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_KEYSTORE_TYPE_CONFIG, "zookeeper.ssl.keyStore.type");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_TRUSTSTORE_LOCATION_CONFIG, "zookeeper.ssl.trustStore.location");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_TRUSTSTORE_PASSWORD_CONFIG, "zookeeper.ssl.trustStore.password");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_TRUSTSTORE_TYPE_CONFIG, "zookeeper.ssl.trustStore.type");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_PROTOCOL_CONFIG, "zookeeper.ssl.protocol");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_ENABLED_PROTOCOLS_CONFIG, "zookeeper.ssl.enabledProtocols");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_CIPHER_SUITES_CONFIG, "zookeeper.ssl.ciphersuites");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM_CONFIG, "zookeeper.ssl.hostnameVerification");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_CRL_ENABLE_CONFIG, "zookeeper.ssl.crl");
+        ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_OCSP_ENABLE_CONFIG, "zookeeper.ssl.ocsp");
+    }
```
Rather than hardcoding config names (e.g. `"zookeeper.ssl.keyStore.location"` and so on), can we refer to relevant config in `KafkaConfig` (e.g. `KafkaConfig.ZkSslKeyStoreLocationProp()` and so on)?

efeg replied:
Here is an example change that refers to the relevant config name in `KafkaConfig` rather than hardcoding the config name (i.e. just as you did in `ExecutorConfig`):

* Current version:
```
...
ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_KEYSTORE_LOCATION_CONFIG, "zookeeper.ssl.keyStore.location");
...
```

* Proposed version:
```
...
ZK_PROPERTIES.put(ExecutorConfig.ZOOKEEPER_SSL_KEYSTORE_LOCATION_CONFIG, KafkaConfig.ZkSslKeyStoreLocationProp());
...
```

Is it possible to update each `put` entry this way?

mimaison replied:
The names on the right are the configurations ZooKeeper uses. This variable `ZK_PROPERTIES` maps Cruise Control configurations (this is what users will enter and I decided to reuse the same names as Kafka) to ZooKeeper configuration that are set on the ZooKeeper client.

Unfortunately, the ZooKeeper names don't seem to be available anywhere as constants. ZooKeeper only has `ZKClientConfig.ZOOKEEPER_CLIENT_CNXN_SOCKET`. Kafka also builds this mapping but it's not exposed either (https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaConfig.scala#L323-L337).

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/ZKConfigUtils.java:
```diff
@@ -0,0 +1,103 @@
+        } else {
+            ZKClientConfig clientConfig = new ZKClientConfig();
+            clientConfig.setProperty(ZKClientConfig.SECURE_CLIENT, "true");
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_CNXN_SOCKET_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_KEYSTORE_LOCATION_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_KEYSTORE_PASSWORD_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_KEYSTORE_TYPE_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_TRUSTSTORE_LOCATION_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_TRUSTSTORE_PASSWORD_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_TRUSTSTORE_TYPE_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_PROTOCOL_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_ENABLED_PROTOCOLS_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_CIPHER_SUITES_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_CRL_ENABLE_CONFIG, config);
+            setZooKeeperClientProperty(clientConfig, ExecutorConfig.ZOOKEEPER_SSL_OCSP_ENABLE_CONFIG, config);
```
Can we simplify this using a loop over the keys of `ZK_PROPERTIES` and calling  `setZooKeeperClientProperty` within the loop? Presumably, this would help with code maintainability -- e.g. if we want to add one more config to `ZK_PROPERTIES` in the future.

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/ExecutorConfig.java:
```diff
@@ -38,6 +38,21 @@
   public static final boolean DEFAULT_ZOOKEEPER_SECURITY_ENABLED = false;
   public static final String ZOOKEEPER_SECURITY_ENABLED_DOC = "Specify if ZooKeeper is secured, true or false";
 
+  public static final String ZOOKEEPER_SSL_CLIENT_ENABLE_CONFIG = "zookeeper.ssl.client.enable";
+  public static final String ZOOKEEPER_SSL_CNXN_SOCKET_CONFIG = "zookeeper.clientCnxnSocket";
+  public static final String ZOOKEEPER_SSL_KEYSTORE_LOCATION_CONFIG = "zookeeper.ssl.keystore.location";
+  public static final String ZOOKEEPER_SSL_KEYSTORE_PASSWORD_CONFIG = "zookeeper.ssl.keystore.password";
+  public static final String ZOOKEEPER_SSL_KEYSTORE_TYPE_CONFIG = "zookeeper.ssl.keystore.type";
+  public static final String ZOOKEEPER_SSL_TRUSTSTORE_LOCATION_CONFIG = "zookeeper.ssl.truststore.location";
+  public static final String ZOOKEEPER_SSL_TRUSTSTORE_PASSWORD_CONFIG = "zookeeper.ssl.truststore.password";
+  public static final String ZOOKEEPER_SSL_TRUSTSTORE_TYPE_CONFIG = "zookeeper.ssl.truststore.type";
+  public static final String ZOOKEEPER_SSL_PROTOCOL_CONFIG = "zookeeper.ssl.protocol";
+  public static final String ZOOKEEPER_SSL_ENABLED_PROTOCOLS_CONFIG = "zookeeper.ssl.enabled.protocols";
+  public static final String ZOOKEEPER_SSL_CIPHER_SUITES_CONFIG = "zookeeper.ssl.cipher.suites";
+  public static final String ZOOKEEPER_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM_CONFIG = "zookeeper.ssl.endpoint.identification.algorithm";
+  public static final String ZOOKEEPER_SSL_CRL_ENABLE_CONFIG = "zookeeper.ssl.crl.enable";
+  public static final String ZOOKEEPER_SSL_OCSP_ENABLE_CONFIG = "zookeeper.ssl.ocsp.enable";
+
```
1. Can we follow the same structure for new configs - i.e. for each config:

```
  /**
   * <code>config.name</code>
   */
  public static final String CONFIG_NAME_CONFIG = "config.name";
  public static final int DEFAULT_CONFIG_NAME = <default-config-value-here>;
  public static final String CONFIG_NAME_DOC = "...";
```

2. Rather than hardcoding config names (e.g. `"zookeeper.ssl.keyStore.location"` and so on), can we refer to relevant config in `KafkaConfig` (e.g. `KafkaConfig.ZkSslKeyStoreLocationProp()` and so on)?

efeg replied:
3. Also, can we update `Configurations.md` with these new configs?

efeg commented on cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControlUtils.java:
```diff
@@ -511,9 +511,13 @@ public static boolean containsAny(Set<Integer> a, Set<Integer> b) {
    * @return A new instance of KafkaZkClient
    */
-  public static KafkaZkClient createKafkaZkClient(String connectString, String metricGroup, String metricType, boolean zkSecurityEnabled) {
+  public static KafkaZkClient createKafkaZkClient(String connectString,
```
Nit: Missing `@param zkClientConfig` in JavaDoc.

efeg commented on cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/executor/ExecutorTest.java:
```diff
@@ -118,7 +118,8 @@ public void testReplicaReassignment() throws InterruptedException, OngoingExecut
                                                                               "ReplicaReassignment",
-                                                                              false);
+                                                                              false,
+                                                                              null);
```
1. [Applies to the updates in `ReplicationThrottleHelperTest`, as well] In the main code, there is no code path that passes `null` to the `KafkaCruiseControlUtils#createKafkaZkClient`. Should we pass the return value of `ZKConfigUtils#zkClientConfigFromKafkaConfig` for some config to make the test more in line with the typical usage?
2. Would it be possible to add some tests to see that the new functionality works as expected (`CCKafkaIntegrationTestHarness` might be useful)?

efeg reviewed (commented):
Thanks for resolving the style issues.
Left some comments.

mimaison commented:
@egyedt @efeg Thanks for the feedback. I've addressed all your comments except for the integration test. I only took a quick look but it seemed like adding SSL to ZooKeeper in the integration test harness is not a small task. I'll try to take another look this week.

efeg reviewed (commented):
@mimaison Thanks for the updates!
Overall, the PR looks good to me -- other than the integration test, I just noticed a minor feedback missed from the earlier review.

mimaison commented:
@efeg I took another look at adding SSL support to `CCEmbeddedZookeeper` and it looks rather tricky. I'm not familiar with these APIs so let me know if you have pointers how to do it. Otherwise I'd be tempted to not include it in this PR. Thanks

egyedt commented:
@efeg, @mimaison 
In this case, I suppose this can be merged if the conflicts will be resolved.

mimaison commented:
@egyedt Thanks, I rebased on `migrate_to_kafka_2_5`.

efeg reviewed (approved):
Thanks for adding support to connect ZooKeeper with TLS!
LGTM.

mimaison commented:
Thanks @efeg and @egyedt for the reviews and feedback! Great to see this merged
